### PR TITLE
Add support for gradient accumulation

### PR DESCRIPTION
**Pull request by romerojosh**
This PR adds support for gradient accumulation, which can be used to increase the effective batch size during training (and reduce the amount of gradient allreduce communication during distributed training). The implementation does not perform any additional loss normalization by the number of accumulation steps so users of this functionality will need to take care to modify learning rates appropriately.

The feature is enabled via a new configuration argument `grad_accumulation_steps` within the `optimizer` block:
```
optimizer:
  type: ...
  general:
    grad_accumulation_steps: 4
```

### CODE REVIEW

File: src/csrc/include/internal/model_state.h [hunk 1]
```diff
@@ -40,6 +40,7 @@ namespace torchfort {
 struct ModelState {
   int64_t step_train;
   int64_t step_inference;
+  int64_t step_train_current; // training step of current run (ignoring restarted state)
```
Comment (azrael417) [inline src/csrc/include/internal/model_state.h:43]: why do we need this additional counter btw? Can step_train not be used in place of step_train_current?

File: src/csrc/training.cpp [hunk 2]
```diff
@@ -129,6 +129,7 @@ void train_multiarg(const char* name, torchfort_tensor_list_t inputs_in, torchfo
 
   model->train();
   auto opt = models[name].optimizer.get();
```
Comment (azrael417) [inline src/csrc/training.cpp:131]: I think we can remove the .get() here and in the line below.

File: tests/supervised/test_training.cpp [hunk 3]
```diff
@@ -270,6 +270,68 @@ void training_test_multiarg_errors(const std::string& model_config) {
...
+    }
+  }
+
+  free_data_ptr(input_ptr, dev_input);
+  free_data_ptr(label_ptr, dev_input);
+  free_data_ptr(output_ptr, dev_input);
+  free_data_ptr(output2_ptr, dev_input);
+
+}
+
```
Comment (azrael417) [inline tests/supervised/test_training.cpp:334]: one other test would be to feed a batch of say size 2 and compare the grads of the model to two batches of size 1 with grad accumulation? The was what I had in mind for the RL stuff, not sure it is easy to implement.

**Review by azrael417 — APPROVED** [review]
LGTM, just a few minor comment. Thanks for adding grad_accumulation_steps to some of the RL model wrapper classes.

File: src/csrc/include/internal/model_state.h (hunk 1 above)
Comment (romerojosh) [reply src/csrc/include/internal/model_state.h:43]: This is handle situations where one might save a checkpoint at a training step that is not a multiple of `grad_accumulation_steps` (e.g. say they save a checkpoint at `step_train = 10` with `grad_accumulation_steps = 4`). Now, if they were to restart, if we used `step_train` for our gradient accumulation timing, we would perform a weight update after 2 steps, which is not correct. `step_train_current` always begins at zero regardless of the restarted state, so we can use that to ensure we perform weight updates at the correct interval.

Comment (azrael417) [reply src/csrc/include/internal/model_state.h:43]: That makes total sense, thank you.

File: tests/supervised/test_training.cpp (hunk 3 above)
Comment (romerojosh) [reply tests/supervised/test_training.cpp:334]: Direct access to the gradients in our tests is difficult. That is why I am comparing inference output in the existing test. I think the current test is sufficient for now.

File: src/csrc/training.cpp (hunk 2 above)
Comment (romerojosh) [reply src/csrc/training.cpp]: Sure, I just removed `.get()` usage everywhere in this file.

File: tests/supervised/test_training.cpp (hunk 3 above)
Comment (azrael417) [reply tests/supervised/test_training.cpp:334]: yes, I agree.

File: src/csrc/training.cpp (hunk 2 above)
Comment (azrael417) [reply src/csrc/training.cpp]: great, thanks. It is correct either way but I think it is cleaner using the smart pointers.
